File: scripts/crawl.mjs
/**
 * Phase 1: crawl taisiy.ru (OpenCart) into data/raw/
 * No invented fields. Missing values stay null / [].
 */
import { mkdir, writeFile } from "node:fs/promises";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, "..");
const RAW = join(ROOT, "data", "raw");
const BASE = "https://taisiy.ru";
const UA = "Mozilla/5.0 (compatible; TaisiyaCatalogAudit/1.0)";

const CATEGORY_URLS = [
  { slug: "katalog", url: `${BASE}/katalog`, name: "Каталог" },
  { slug: "kostyumy", url: `${BASE}/katalog/kostyumy`, name: "Костюмы" },
  { slug: "pizhamy", url: `${BASE}/katalog/pizhamy`, name: "Пижамы для дома" },
  { slug: "platjya", url: `${BASE}/katalog/platjya`, name: "Платья" },
  { slug: "sarafany", url: `${BASE}/katalog/sarafany`, name: "Сарафаны" },
  { slug: "tolstovki", url: `${BASE}/katalog/tolstovki`, name: "Толстовки" },
  { slug: "tuniki", url: `${BASE}/katalog/tuniki`, name: "Туники" },
  { slug: "vodolazki", url: `${BASE}/vodolazki`, name: "Водолазки" },
  { slug: "sorochki", url: `${BASE}/sorochki`, name: "Сорочки" },
  { slug: "futbolki", url: `${BASE}/futbolki`, name: "Футболки" },
  { slug: "khalaty", url: `${BASE}/khalaty`, name: "Халаты" },
  { slug: "velosipedki", url: `${BASE}/velosipedki`, name: "Велосипедки" },
  { slug: "topy", url: `${BASE}/topy`, name: "Топы" },
];

const INFO_PAGES = [
  { id: "home", url: `${BASE}/` },
  { id: "about", url: `${BASE}/about_us` },
  { id: "delivery", url: `${BASE}/delivery` },
  { id: "privacy", url: `${BASE}/privacy` },
  { id: "size-guide", url: `${BASE}/tablica-razmerov` },
  { id: "contact", url: `${BASE}/contact` },
  { id: "sitemap", url: `${BASE}/sitemap` },
  { id: "return", url: `${BASE}/return-add` },
  { id: "search", url: `${BASE}/index.php?route=product/search&search=&limit=100` },
];

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function fetchText(url) {
  const res = await fetch(url, { headers: { "User-Agent": UA, Accept: "text/html" } });
  const html = await res.text();
  return { status: res.status, url: res.url, html };
}

function decodeHtml(s) {
  return s
    .replace(/&nbsp;/g, " ")
    .replace(/&amp;/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/\s+/g, " ")
    .trim();
}

function stripTags(html) {
  return decodeHtml(html.replace(/<br\s*\/?>/gi, "\n").replace(/<[^>]+>/g, " "));
}

function cacheToOriginal(url) {
  if (!url) return null;
  let u = url.split("?")[0];
  u = u.replace("/image/cache/", "/image/");
  u = u.replace(/-(\d+)x(\d+)(\.(jpe?g|png|webp|gif))$/i, "$3");
  return u;
}

function unique(arr) {
  return [...new Set(arr.filter(Boolean))];
}

function extractProductLinks(html) {
  const links = [];
  const re = /<div class="product-thumb">[\s\S]*?<h4><a href="([^"]+)">([\s\S]*?)<\/a><\/h4>/g;
  let m;
  while ((m = re.exec(html))) {
    const href = m[1].replace(/&amp;/g, "&").split("?")[0];
    links.push({ url: href, name: decodeHtml(m[2]) });
  }
  return links;
}

function extractPaginationPages(html, baseUrl) {
  const pages = new Set([baseUrl]);
  const re = /href="(https:\/\/taisiy\.ru\/[^"]*(?:[?&]page=\d+)[^"]*)"/g;
  let m;
  while ((m = re.exec(html))) {
    pages.add(m[1].replace(/&amp;/g, "&"));
  }
  const shown = html.match(/Показано с \d+ по \d+ из (\d+)/);
  const total = shown ? Number(shown[1]) : null;
  return { pages: [...pages], total };
}

function parseProduct(html, url) {
  const nameMatch = html.match(/<h1[^>]*>([\s\S]*?)<\/h1>/);
  const name = nameMatch ? decodeHtml(nameMatch[1]) : null;

  const articleMatch = html.match(/<li><b>Модель:<\/b>\s*([^<]+)<\/li>/);
  const article = articleMatch ? decodeHtml(articleMatch[1]) : null;

  const availabilityMatch = html.match(/<li><b>Наличие:<\/b>\s*([^<]+)<\/li>/);
  const availability = availabilityMatch ? decodeHtml(availabilityMatch[1]) : null;

  const idMatch = html.match(/wishlist\.add\('(\d+)'\)/);
  const ocId = idMatch ? idMatch[1] : null;

  const retailMatch = html.match(/data-price="([\d.]+)"\s+class="calc-price"/);
  const priceRetail = retailMatch ? Number(retailMatch[1]) : null;

  const wholesaleMatch = html.match(/<span class="opt-chena">([^<]+)<\/span>/);
  let priceWholesale = null;
  if (wholesaleMatch) {
    const n = wholesaleMatch[1].replace(/[^\d.,]/g, "").replace(",", ".");
    priceWholesale = n ? Number(n) : null;
  }

  const sizes = [];
  const sizeBlock = html.match(/<label class="control-label">Размер<\/label>([\s\S]*?)<\/div>\s*<\/div>/);
  if (sizeBlock) {
    const sizeRe = /<label for="\d+">\s*([^<]+?)\s*<\/label>/g;
    let sm;
    while ((sm = sizeRe.exec(sizeBlock[1]))) {
      const s = decodeHtml(sm[1]);
      if (s) sizes.push(s);
    }
  }

  const descMatch = html.match(/id="tab-description">([\s\S]*?)<\/div>/);
  const description = descMatch ? stripTags(descMatch[1]).replace(/\n+/g, "\n").trim() : null;

  const material = extractMaterial(description);

  const imageUrls = [];
  const imgRe = /<a class="thumbnail" href="(https:\/\/taisiy\.ru\/image\/cache\/[^"]+)"/g;
  let im;
  while ((im = imgRe.exec(html))) {
    imageUrls.push(im[1]);
  }

  const breadcrumbs = [];
  const bcRe = /<ul class="breadcrumb">[\s\S]*?<\/ul>/;
  const bc = html.match(bcRe);
  if (bc) {
    const aRe = /<a href="[^"]*">([\s\S]*?)<\/a>/g;
    let bm;
    while ((bm = aRe.exec(bc[0]))) {
      breadcrumbs.push(decodeHtml(bm[1].replace(/<[^>]+>/g, "")));
    }
  }

  return {
    sourceUrl: url,
    ocId,
    name,
    article,
    availability,
    priceRetail,
    priceWholesale,
    sizes: unique(sizes),
    colors: [],
    material,
    description: description || null,
    imageCacheUrls: unique(imageUrls),
    imageOriginalCandidates: unique(imageUrls.map(cacheToOriginal)),
    breadcrumbs,
  };
}

function extractMaterial(description) {
  if (!description) return null;
  const m =
    description.match(/ткань\s+100%\s*хлопок/i) ||
    description.match(/из\s+100%\s*хлопка/i) ||
    description.match(/100%\s*хлопок/i) ||
    description.match(/из\s+натурального\s+хлопка/i) ||
    description.match(/из\s+кулирки/i);
  if (!m) return null;
  const t = m[0].toLowerCase();
  if (t.includes("кулирк")) return "кулирка";
  if (t.includes("хлоп")) return "100% хлопок";
  return m[0];
}

function parseSizeTable(html) {
  const rows = [];
  const table = html.match(/<table[\s\S]*?<\/table>/i);
  if (!table) return rows;
  const trRe = /<tr[\s\S]*?<\/tr>/gi;
  const trs = table[0].match(trRe) || [];
  for (const tr of trs) {
    const cells = [...tr.matchAll(/<t[dh][^>]*>([\s\S]*?)<\/t[dh]>/gi)].map((c) =>
      decodeHtml(c[1].replace(/<[^>]+>/g, " "))
    );
    if (cells.length) rows.push(cells);
  }
  return rows;
}

function parseCategoryFilters(html) {
  const filters = { sizes: [], materials: [], colors: [], manufacturers: [], priceRange: null };
  const sizeSec = html.match(/Размер[\s\S]*?(?=<div class="filter-group"|Материал|Производитель|Цвет|Сбросить|$)/);
  // Simpler: grab filter labels near known headings
  const colorBlock = html.match(/>Цвет<\/[a-z]+>([\s\S]*?)(?:Сбросить|<\/form>)/i);
  if (colorBlock) {
    const names = colorBlock[1].matchAll(/>([а-яёa-z\- ]+)\s+\d+\s*</gi);
    for (const n of names) filters.colors.push(decodeHtml(n[1]));
  }
  const matBlock = html.match(/>Материал<\/[a-z]+>([\s\S]*?)(?:Производитель|Цвет|Сбросить)/i);
  if (matBlock) {
    const names = matBlock[1].matchAll(/>([^<]+?)\s+\d+\s*</g);
    for (const n of names) filters.materials.push(decodeHtml(n[1]));
  }
  return filters;
}

function extractCategoryNav(html) {
  const cats = [];
  const re = /<a href="(https:\/\/taisiy\.ru\/katalog\/[^"]+)"[^>]*>[\s\-]*([^<]+)</g;
  let m;
  while ((m = re.exec(html))) {
    cats.push({ url: m[1], name: decodeHtml(m[2]) });
  }
  const extra = html.matchAll(/<li><a href="(https:\/\/taisiy\.ru\/[^"]+)">\s*-\s*([^<]+)</g);
  for (const e of extra) {
    cats.push({ url: e[1], name: decodeHtml(e[2]) });
  }
  const seen = new Set();
  return cats.filter((c) => {
    if (seen.has(c.url)) return false;
    seen.add(c.url);
    return true;
  });
}

async function crawlCategory(cat) {
  const first = await fetchText(cat.url);
  await mkdir(join(RAW, "categories"), { recursive: true });
  await writeFile(join(RAW, "categories", `${cat.slug}.html`), first.html, "utf8");

  if (first.status >= 400 || /не найдена|нет товаров/i.test(first.html)) {
    const empty = /нет товаров/i.test(first.html);
    return {
      ...cat,
      status: first.status,
      empty: empty || first.status >= 400,
      productCountListed: 0,
      products: [],
      pagesFetched: 1,
    };
  }

  const { pages, total } = extractPaginationPages(first.html, cat.url);
  let allHtml = first.html;
  const extraPages = pages.filter((p) => p !== cat.url && /[?&]page=\d+/.test(p));
  extraPages.sort((a, b) => {
    const pa = Number((a.match(/[?&]page=(\d+)/) || [])[1] || 0);
    const pb = Number((b.match(/[?&]page=(\d+)/) || [])[1] || 0);
    return pa - pb;
  });

  for (const p of extraPages) {
    await sleep(250);
    const next = await fetchText(p);
    const pageNum = (p.match(/[?&]page=(\d+)/) || [])[1] || "x";
    await writeFile(join(RAW, "categories", `${cat.slug}-p${pageNum}.html`), next.html, "utf8");
    allHtml += "\n" + next.html;
  }

  const products = extractProductLinks(allHtml);
  const seen = new Set();
  const uniq = products.filter((p) => {
    if (seen.has(p.url)) return false;
    seen.add(p.url);
    return true;
  });

  return {
    ...cat,
    status: first.status,
    empty: uniq.length === 0,
    productCountListed: total,
    productCountFound: uniq.length,
    products: uniq,
    pagesFetched: 1 + extraPages.length,
    nav: extractCategoryNav(first.html),
  };
}

async function main() {
  await mkdir(RAW, { recursive: true });
  await mkdir(join(RAW, "products"), { recursive: true });
  await mkdir(join(RAW, "pages"), { recursive: true });

  const pages = {};
  for (const p of INFO_PAGES) {
    console.log("page", p.id, p.url);
    const r = await fetchText(p.url);
    await writeFile(join(RAW, "pages", `${p.id}.html`), r.html, "utf8");
    pages[p.id] = { url: p.url, status: r.status, finalUrl: r.url, bytes: r.html.length };
    await sleep(200);
  }

  const categories = [];
  const productIndex = new Map();

  for (const cat of CATEGORY_URLS) {
    console.log("category", cat.slug);
    const crawled = await crawlCategory(cat);
    categories.push(crawled);
    for (const p of crawled.products || []) {
      if (!productIndex.has(p.url)) {
        productIndex.set(p.url, { url: p.url, name: p.name, categories: [crawled.name] });
      } else {
        const rec = productIndex.get(p.url);
        if (!rec.categories.includes(crawled.name)) rec.categories.push(crawled.name);
      }
    }
    await sleep(250);
  }

  // Search results as completeness source
  const searchHtml = await (await import("node:fs/promises")).readFile(join(RAW, "pages", "search.html"), "utf8");
  for (const p of extractProductLinks(searchHtml)) {
    if (!productIndex.has(p.url)) {
      productIndex.set(p.url, { url: p.url, name: p.name, categories: ["(только поиск)"] });
    }
  }

  const products = [];
  let i = 0;
  for (const rec of productIndex.values()) {
    i += 1;
    console.log(`product ${i}/${productIndex.size}`, rec.url);
    const r = await fetchText(rec.url);
    const slug = rec.url.replace(BASE + "/", "").replace(/[^\w\-а-яё]+/gi, "_") || `p${i}`;
    await writeFile(join(RAW, "products", `${slug}.html`), r.html, "utf8");
    const parsed = parseProduct(r.html, rec.url);
    parsed.listedName = rec.name;
    parsed.listedCategories = rec.categories.filter((c) => c !== "Каталог");
    products.push(parsed);
    await sleep(200);
  }

  const sizeHtml = await (await import("node:fs/promises")).readFile(join(RAW, "pages", "size-guide.html"), "utf8");
  const aboutHtml = await (await import("node:fs/promises")).readFile(join(RAW, "pages", "about.html"), "utf8");
  const deliveryHtml = await (await import("node:fs/promises")).readFile(join(RAW, "pages", "delivery.html"), "utf8");
  const privacyHtml = await (await import("node:fs/promises")).readFile(join(RAW, "pages", "privacy.html"), "utf8");
  const contactHtml = await (await import("node:fs/promises")).readFile(join(RAW, "pages", "contact.html"), "utf8");

  const dump = {
    crawledAt: new Date().toISOString(),
    source: BASE,
    pages,
    categories: categories.map((c) => ({
      slug: c.slug,
      name: c.name,
      url: c.url,
      status: c.status,
      empty: c.empty,
      productCountListed: c.productCountListed,
      productCountFound: c.productCountFound,
      pagesFetched: c.pagesFetched,
      products: c.products,
    })),
    products,
    sizeTable: parseSizeTable(sizeHtml),
    aboutText: stripTags((aboutHtml.match(/id="content"[\s\S]*?<h1[\s\S]*?<\/h1>([\s\S]*?)<footer/) || [])[1] || ""),
    deliveryText: stripTags((deliveryHtml.match(/id="content"[\s\S]*?<h1[\s\S]*?<\/h1>([\s\S]*?)<footer/) || [])[1] || ""),
    privacyText: stripTags((privacyHtml.match(/id="content"[\s\S]*?<h1[\s\S]*?<\/h1>([\s\S]*?)<footer/) || [])[1] || ""),
    contactHtmlExcerpt: stripTags((contactHtml.match(/id="content"[\s\S]*?<h1[\s\S]*?<\/h1>([\s\S]*?)Форма обратной связи/) || [])[1] || ""),
  };

  await writeFile(join(RAW, "crawl-dump.json"), JSON.stringify(dump, null, 2), "utf8");
  console.log("DONE products:", products.length, "categories:", categories.length);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/crawl.py
# -*- coding: utf-8 -*-
"""Phase 1: crawl taisiy.ru (OpenCart) into data/raw/."""
from __future__ import annotations

import json
import re
import time
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime, timezone
from html import unescape
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
RAW = ROOT / "data" / "raw"
BASE = "https://taisiy.ru"
UA = "Mozilla/5.0 (compatible; TaisiyaCatalogAudit/1.0)"

CATEGORY_URLS = [
    {"slug": "katalog", "url": f"{BASE}/katalog", "name": "Каталог"},
    {"slug": "kostyumy", "url": f"{BASE}/katalog/kostyumy", "name": "Костюмы"},
    {"slug": "pizhamy", "url": f"{BASE}/katalog/pizhamy", "name": "Пижамы для дома"},
    {"slug": "platjya", "url": f"{BASE}/katalog/platjya", "name": "Платья"},
    {"slug": "sarafany", "url": f"{BASE}/katalog/sarafany", "name": "Сарафаны"},
    {"slug": "tolstovki", "url": f"{BASE}/katalog/tolstovki", "name": "Толстовки"},
    {"slug": "tuniki", "url": f"{BASE}/katalog/tuniki", "name": "Туники"},
    {"slug": "vodolazki", "url": f"{BASE}/vodolazki", "name": "Водолазки"},
    {"slug": "sorochki", "url": f"{BASE}/sorochki", "name": "Сорочки"},
    {"slug": "futbolki", "url": f"{BASE}/futbolki", "name": "Футболки"},
    {"slug": "khalaty", "url": f"{BASE}/khalaty", "name": "Халаты"},
    {"slug": "velosipedki", "url": f"{BASE}/velosipedki", "name": "Велосипедки"},
    {"slug": "topy", "url": f"{BASE}/topy", "name": "Топы"},
]

INFO_PAGES = [
    {"id": "home", "url": f"{BASE}/"},
    {"id": "about", "url": f"{BASE}/about_us"},
    {"id": "delivery", "url": f"{BASE}/delivery"},
    {"id": "privacy", "url": f"{BASE}/privacy"},
    {"id": "size-guide", "url": f"{BASE}/tablica-razmerov"},
    {"id": "contact", "url": f"{BASE}/contact"},
    {"id": "sitemap", "url": f"{BASE}/sitemap"},
    {"id": "return", "url": f"{BASE}/return-add"},
    {"id": "search", "url": f"{BASE}/index.php?route=product/search&search=&limit=100"},
]


def encode_url(url: str) -> str:
    parts = urllib.parse.urlsplit(url)
    path = urllib.parse.quote(parts.path, safe="/")
    query = urllib.parse.quote(parts.query, safe="=&%")
    return urllib.parse.urlunsplit((parts.scheme, parts.netloc, path, query, parts.fragment))


def fetch_text(url: str, retries: int = 4) -> tuple[int, str, str]:
    req = urllib.request.Request(
        encode_url(url), headers={"User-Agent": UA, "Accept": "text/html"}
    )
    last_err: Exception | None = None
    for attempt in range(retries):
        try:
            with urllib.request.urlopen(req, timeout=40) as res:
                html = res.read().decode("utf-8", errors="replace")
                return res.status, res.geturl(), html
        except urllib.error.HTTPError as e:
            html = e.read().decode("utf-8", errors="replace") if e.fp else ""
            return e.code, url, html
        except Exception as e:
            last_err = e
            time.sleep(1.2 * (attempt + 1))
    raise last_err  # type: ignore[misc]


def product_key(url: str) -> str:
    path = urllib.parse.urlsplit(url).path.rstrip("/")
    slug = path.split("/")[-1]
    return urllib.parse.unquote(slug).lower()


def decode_html(s: str) -> str:
    s = unescape(s)
    s = s.replace("\xa0", " ")
    s = re.sub(r"\s+", " ", s).strip()
    return s


def strip_tags(html: str) -> str:
    html = re.sub(r"<br\s*/?>", "\n", html, flags=re.I)
    html = re.sub(r"<[^>]+>", " ", html)
    return decode_html(html).replace(" \n", "\n")


def unique(items: list) -> list:
    out = []
    seen = set()
    for x in items:
        if not x or x in seen:
            continue
        seen.add(x)
        out.append(x)
    return out


def cache_to_original(url: str | None) -> str | None:
    if not url:
        return None
    u = url.split("?")[0]
    u = u.replace("/image/cache/", "/image/")
    u = re.sub(r"-(\d+)x(\d+)(\.(jpe?g|png|webp|gif))$", r"\3", u, flags=re.I)
    return u


def extract_product_links(html: str) -> list[dict]:
    links = []
    for m in re.finditer(
        r'<div class="product-thumb">[\s\S]*?<h4><a href="([^"]+)">([\s\S]*?)</a></h4>',
        html,
    ):
        href = unescape(m.group(1)).split("?")[0]
        links.append({"url": href, "name": decode_html(m.group(2))})
    return links


def extract_pagination_pages(html: str, base_url: str) -> tuple[list[str], int | None]:
    pages = {base_url}
    for m in re.finditer(r'href="(https://taisiy\.ru/[^"]*(?:[?&]page=\d+)[^"]*)"', html):
        pages.add(unescape(m.group(1)))
    shown = re.search(r"Показано с \d+ по \d+ из (\d+)", html)
    total = int(shown.group(1)) if shown else None
    return list(pages), total


def extract_material(description: str | None) -> str | None:
    if not description:
        return None
    patterns = [
        r"ткань\s+100%\s*хлопок",
        r"из\s+100%\s*хлопка",
        r"100%\s*хлопок",
        r"из\s+натурального\s+хлопка",
        r"из\s+кулирки",
    ]
    for p in patterns:
        m = re.search(p, description, flags=re.I)
        if m:
            t = m.group(0).lower()
            if "кулирк" in t:
                return "кулирка"
            if "хлоп" in t:
                return "100% хлопок"
            return m.group(0)
    return None


def parse_product(html: str, url: str) -> dict:
    name_m = re.search(r"<h1[^>]*>([\s\S]*?)</h1>", html)
    name = decode_html(name_m.group(1)) if name_m else None

    article_m = re.search(r"<li><b>Модель:</b>\s*([^<]+)</li>", html)
    article = decode_html(article_m.group(1)) if article_m else None

    avail_m = re.search(r"<li><b>Наличие:</b>\s*([^<]+)</li>", html)
    availability = decode_html(avail_m.group(1)) if avail_m else None

    id_m = re.search(r"wishlist\.add\('(\d+)'\)", html)
    oc_id = id_m.group(1) if id_m else None

    retail_m = re.search(r'data-price="([\d.]+)"\s+class="calc-price"', html)
    price_retail = float(retail_m.group(1)) if retail_m else None

    wholesale_m = re.search(r'<span class="opt-chena">([^<]+)</span>', html)
    price_wholesale = None
    if wholesale_m:
        n = re.sub(r"[^\d.,]", "", wholesale_m.group(1)).replace(",", ".")
        price_wholesale = float(n) if n else None

    sizes = []
    size_block = re.search(
        r'<label class="control-label">Размер</label>([\s\S]*?)</div>\s*</div>', html
    )
    if size_block:
        for sm in re.finditer(r'<label for="\d+">\s*([^<]+?)\s*</label>', size_block.group(1)):
            s = decode_html(sm.group(1))
            if s:
                sizes.append(s)

    desc_m = re.search(r'id="tab-description">([\s\S]*?)</div>', html)
    description = strip_tags(desc_m.group(1)) if desc_m else None
    if description:
        description = re.sub(r"\n+", "\n", description).strip() or None

    image_urls = re.findall(r'<a class="thumbnail" href="(https://taisiy\.ru/image/cache/[^"]+)"', html)

    return {
        "sourceUrl": url,
        "ocId": oc_id,
        "name": name,
        "article": article,
        "availability": availability,
        "priceRetail": price_retail,
        "priceWholesale": price_wholesale,
        "sizes": unique(sizes),
        "colors": [],
        "material": extract_material(description),
        "description": description,
        "imageCacheUrls": unique(image_urls),
        "imageOriginalCandidates": unique([cache_to_original(u) for u in image_urls]),
    }


def parse_size_table(html: str) -> list[list[str]]:
    rows = []
    table = re.search(r"<table[\s\S]*?</table>", html, flags=re.I)
    if not table:
        return rows
    for tr in re.findall(r"<tr[\s\S]*?</tr>", table.group(0), flags=re.I):
        cells = [
            decode_html(re.sub(r"<[^>]+>", " ", c))
            for c in re.findall(r"<t[dh][^>]*>([\s\S]*?)</t[dh]>", tr, flags=re.I)
        ]
        if cells:
            rows.append(cells)
    return rows


def extract_h1_content(html: str) -> str:
    m = re.search(
        r'<div id="content"[^>]*>[\s\S]*?<h1[^>]*>[\s\S]*?</h1>([\s\S]*?)</div>\s*<footer',
        html,
        flags=re.I,
    )
    if not m:
        m = re.search(r"<h1[^>]*>[\s\S]*?</h1>([\s\S]*?)<footer", html, flags=re.I)
    return strip_tags(m.group(1)) if m else ""


def slug_from_url(url: str, i: int) -> str:
    slug = url.replace(BASE + "/", "")
    slug = re.sub(r"[^\w\-]+", "_", slug, flags=re.U)
    return slug or f"p{i}"


def crawl_category(cat: dict) -> dict:
    status, final_url, html = fetch_text(cat["url"])
    (RAW / "categories").mkdir(parents=True, exist_ok=True)
    (RAW / "categories" / f"{cat['slug']}.html").write_text(html, encoding="utf-8")

    not_found = status >= 400 or bool(re.search(r"не найдена", html, flags=re.I))
    empty = bool(re.search(r"нет товаров", html, flags=re.I))
    if not_found or empty:
        return {
            **cat,
            "status": status,
            "empty": True,
            "productCountListed": 0,
            "productCountFound": 0,
            "products": [],
            "pagesFetched": 1,
        }

    pages, total = extract_pagination_pages(html, cat["url"])
    extra = [p for p in pages if p != cat["url"] and re.search(r"[?&]page=\d+", p)]
    extra.sort(key=lambda p: int(re.search(r"[?&]page=(\d+)", p).group(1)))
    all_html = html
    for p in extra:
        time.sleep(0.25)
        _, _, next_html = fetch_text(p)
        page_num = re.search(r"[?&]page=(\d+)", p).group(1)
        (RAW / "categories" / f"{cat['slug']}-p{page_num}.html").write_text(next_html, encoding="utf-8")
        all_html += "\n" + next_html

    products = extract_product_links(all_html)
    seen = set()
    uniq = []
    for p in products:
        if p["url"] in seen:
            continue
        seen.add(p["url"])
        uniq.append(p)

    return {
        **cat,
        "status": status,
        "empty": len(uniq) == 0,
        "productCountListed": total,
        "productCountFound": len(uniq),
        "products": uniq,
        "pagesFetched": 1 + len(extra),
    }


def main() -> None:
    RAW.mkdir(parents=True, exist_ok=True)
    (RAW / "products").mkdir(parents=True, exist_ok=True)
    (RAW / "pages").mkdir(parents=True, exist_ok=True)
    (RAW / "categories").mkdir(parents=True, exist_ok=True)

    pages = {}
    for p in INFO_PAGES:
        print("page", p["id"], p["url"], flush=True)
        status, final_url, html = fetch_text(p["url"])
        (RAW / "pages" / f"{p['id']}.html").write_text(html, encoding="utf-8")
        pages[p["id"]] = {
            "url": p["url"],
            "status": status,
            "finalUrl": final_url,
            "bytes": len(html.encode("utf-8")),
        }
        time.sleep(0.2)

    categories = []
    product_index: dict[str, dict] = {}

    def add_product(url: str, name: str, category: str | None) -> None:
        key = product_key(url)
        if not key:
            return
        if key not in product_index:
            product_index[key] = {
                "url": url.split("?")[0],
                "name": name,
                "categories": [category] if category else [],
            }
            return
        rec = product_index[key]
        if category and category not in rec["categories"]:
            rec["categories"].append(category)
        # Prefer a shorter canonical URL when available
        if len(url.split("?")[0]) < len(rec["url"]):
            rec["url"] = url.split("?")[0]

    for cat in CATEGORY_URLS:
        print("category", cat["slug"], flush=True)
        crawled = crawl_category(cat)
        categories.append(crawled)
        for p in crawled.get("products") or []:
            add_product(p["url"], p["name"], crawled["name"])
        time.sleep(0.25)

    search_html = (RAW / "pages" / "search.html").read_text(encoding="utf-8")
    for p in extract_product_links(search_html):
        add_product(p["url"], p["name"], None)

    products = []
    for i, rec in enumerate(product_index.values(), start=1):
        print(f"product {i}/{len(product_index)} {rec['url']}", flush=True)
        slug = slug_from_url(rec["url"], i)
        html_path = RAW / "products" / f"{slug}.html"
        if html_path.exists() and html_path.stat().st_size > 1000:
            html = html_path.read_text(encoding="utf-8")
            status = 200
        else:
            status, _, html = fetch_text(rec["url"])
            html_path.write_text(html, encoding="utf-8")
        parsed = parse_product(html, rec["url"])
        parsed["httpStatus"] = status
        parsed["listedName"] = rec["name"]
        parsed["listedCategories"] = [c for c in rec["categories"] if c != "Каталог"]
        products.append(parsed)
        time.sleep(0.2)

    dump = {
        "crawledAt": datetime.now(timezone.utc).isoformat(),
        "source": BASE,
        "pages": pages,
        "categories": [
            {
                "slug": c["slug"],
                "name": c["name"],
                "url": c["url"],
                "status": c["status"],
                "empty": c["empty"],
                "productCountListed": c.get("productCountListed"),
                "productCountFound": c.get("productCountFound"),
                "pagesFetched": c.get("pagesFetched"),
                "products": c.get("products"),
            }
            for c in categories
        ],
        "products": products,
        "sizeTable": parse_size_table((RAW / "pages" / "size-guide.html").read_text(encoding="utf-8")),
        "aboutText": extract_h1_content((RAW / "pages" / "about.html").read_text(encoding="utf-8")),
        "deliveryText": extract_h1_content((RAW / "pages" / "delivery.html").read_text(encoding="utf-8")),
        "privacyText": extract_h1_content((RAW / "pages" / "privacy.html").read_text(encoding="utf-8")),
        "contactText": extract_h1_content((RAW / "pages" / "contact.html").read_text(encoding="utf-8")),
    }
    (RAW / "crawl-dump.json").write_text(
        json.dumps(dump, ensure_ascii=False, indent=2), encoding="utf-8"
    )
    print("DONE products:", len(products), "categories:", len(categories), flush=True)


if __name__ == "__main__":
    main()
